### PR TITLE
Add com.moddb.TotalChaosRetro

### PR DESCRIPTION
**Pull request by mbugni**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [ ] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

The only patch I applied is at line 84 of the manifest, it is necessary to avoid crashes on start. I didn't submit it to upstream bacause EOL:
https://www.moddb.com/mods/total-chaos/news/end-of-the-line5

NOTE: basically this app is the same of com.moddb.TotalChaos. It differs because:
- the "retro" version (for lower graphics)
- the legacy gzdoom engine (for old computers)
So basically it uses less resources and is more compatible with old hardware. Like TotalChaos, this app need a place to download the .zip source file (see manifest at line 71 and [1]).

[1] https://github.com/flathub/flathub/pull/708

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.moddb.TotalChaosRetro.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.moddb.TotalChaosRetro</id>
+  <launchable type="desktop-id">com.moddb.TotalChaosRetro.desktop</launchable>
+  <name>Total Chaos Retro</name>
```
Comment (AsciiWolf) [inline com.moddb.TotalChaosRetro.appdata.xml:5]: I would personally use the full name in AppData:
```suggestion
  <name>Total Chaos Retro Edition</name>
```
Desktop file name can be just "Total Chaos Retro".

File: com.moddb.TotalChaosRetro.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        url: https://github.com/freedoom/freedoom/releases/download/v0.12.1/freedoom-0.12.1.zip
+        sha256: f42c6810fc89b0282de1466c2c9c7c9818031a8d556256a6db1b69f6a77b5806
+    build-commands:
+      - install -Dm 644 freedoom2.wad /app/share/games/doom
+
+  - name: totalchaos-retro
+    buildsystem: simple
+    sources:
+      - type: archive
+        path: totalchaos_dc_retro_b5.zip
```
Comment (hfiguiere) [inline com.moddb.TotalChaosRetro.yaml:72]: this won't work. Because
1. the archive isn't here
2. the archive SHOULD NOT be here either.

File: com.moddb.TotalChaosRetro.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      - type: file
+        path: com.moddb.TotalChaosRetro.desktop
+      - type: file
+        path: com.moddb.TotalChaosRetro.png
+    build-commands:
+      # Patch shaders to avoid GZDoom runtime errors
+      - unzip totalchaos.pk3 SHADERS/old.fp
+      - sed -i 's/texture2D/texture/g' SHADERS/old.fp
+      - zip -u totalchaos.pk3 SHADERS/old.fp
+      - rm SHADERS -rf
```
Comment (hfiguiere) [inline com.moddb.TotalChaosRetro.yaml:88]:
```suggestion
```

File: com.moddb.TotalChaosRetro.yaml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      - install -Dm 644 freedoom2.wad /app/share/games/doom
+
+  - name: totalchaos-retro
+    buildsystem: simple
+    sources:
+      - type: archive
+        path: totalchaos_dc_retro_b5.zip
+        sha256: 0d8a87edc4fc4ab84495b23bfe655742721195a094dbe604072183842f84336e
+#      - type: archive
+#        url: https://dl.flathub.org/sources/com.moddb.TotalChaosRetro/totalchaos_dc_retro_b5.zip
```
Comment (hfiguiere) [inline com.moddb.TotalChaosRetro.yaml:75]: neither this.

File: com.moddb.TotalChaosRetro.yaml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  - --env=DOOMWADDIR=/app/share/games/doom
+  - --persist=.config/gzdoom
+cleanup:
+  - /include
+  - /lib/*.a
+  - /lib/*.la
+  - /lib/pkgconfig
+  - /share/man
+modules:
+  - name: fluidsynth
```
Comment (hfiguiere) [inline com.moddb.TotalChaosRetro.yaml:21]: you should use the shared-module instead.

**Review by hfiguiere — CHANGES_REQUESTED** [review]
this cannot build.

File: com.moddb.TotalChaosRetro.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      Survive in 6 chapters, against over 8 horrifying fiends with a large assortment of weapons.
+    </p>
+    <p>
+      Retro Edition, has reduced texture resolutions and poly-counts for improved performance.
+      Save files from other versions of Total Chaos are not compatible.
+    </p>
+  </description>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://media.moddb.com/images/downloads/1/173/172310/retro_banner.jpg</image>
```
Comment (AsciiWolf) [inline com.moddb.TotalChaosRetro.appdata.xml:27]: Here is a better screenshot:
```suggestion
      <image>https://media.moddb.com/images/mods/1/26/25250/RETRO3.jpg</image>
```

File: com.moddb.TotalChaosRetro.yaml (hunk 4 above)
Comment (AsciiWolf) [reply com.moddb.TotalChaosRetro.yaml:75]: The problem with Total Chaos Retro is the same as with my Total Chaos Flatpak - files from Mod DB cannot be downloaded automatically. This was resolved by hosting the archive directly on Flathub servers.

This is the archive file for Total Chaos Retro Edition: https://www.moddb.com/mods/total-chaos/downloads/total-chaos-directors-cut-retro-edition-140

File: com.moddb.TotalChaosRetro.yaml (hunk 5 above)
Comment (AsciiWolf) [reply com.moddb.TotalChaosRetro.yaml:21]: Good idea, I should probably also switch to using [the shared-module](https://github.com/flathub/shared-modules/blob/master/linux-audio/fluidsynth2.json) for fluidsynth in my Total Chaos Flatpak. :-)

File: com.moddb.TotalChaosRetro.appdata.xml (hunk 1 above)
Comment (AsciiWolf) [reply com.moddb.TotalChaosRetro.appdata.xml:5]: But this is up to you. Both variants are correct.

File: com.moddb.TotalChaosRetro.appdata.xml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.moddb.TotalChaosRetro</id>
+  <launchable type="desktop-id">com.moddb.TotalChaosRetro.desktop</launchable>
+  <name>Total Chaos Retro</name>
+  <summary>Survival horror set on a remote island known as Fort Oasis</summary>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>CC-BY-NC-SA-3.0</project_license>
+  <developer_name>wadaholic</developer_name>
+  <url type="homepage">https://www.moddb.com/mods/total-chaos/downloads/total-chaos-directors-cut-retro-edition-140</url>
```
Comment (AsciiWolf) [inline com.moddb.TotalChaosRetro.appdata.xml:10]:
```suggestion
  <url type="homepage">https://www.moddb.com/mods/total-chaos</url>
```

File: com.moddb.TotalChaosRetro.yaml (hunk 4 above)
Comment (mbugni) [reply com.moddb.TotalChaosRetro.yaml:75]: Yes, as I mentioned in my PR I need a place for downloading the file like @AsciiWolf suggested.

File: com.moddb.TotalChaosRetro.appdata.xml (hunk 6 above)
Comment (mbugni) [reply com.moddb.TotalChaosRetro.appdata.xml:27]: Yes, it's better.

File: com.moddb.TotalChaosRetro.yaml (hunk 5 above)
Comment (mbugni) [reply com.moddb.TotalChaosRetro.yaml:21]: @AsciiWolf can you help me to do this?

Comment (AsciiWolf) [reply com.moddb.TotalChaosRetro.yaml:21]: Just type:
`git submodule add https://github.com/flathub/shared-modules.git`

and replace the fluidsynth module with:
`  - shared-modules/linux-audio/fluidsynth2.json`

Then git commit with `-am`.

File: com.moddb.TotalChaosRetro.yaml (hunk 4 above)
Comment (mbugni) [reply com.moddb.TotalChaosRetro.yaml]: @hfiguiere I uploaded totalchaos_dc_retro_b5.zip file on github and modified the manifest accordingly.

File: com.moddb.TotalChaosRetro.yaml (hunk 5 above)
Comment (AsciiWolf) [reply com.moddb.TotalChaosRetro.yaml:21]: Just type:
`git submodule add https://github.com/flathub/shared-modules.git`

and replace the fluidsynth module with:
`  - shared-modules/linux-audio/fluidsynth2.json`

Then git commit with `-am`.

**Comment by AsciiWolf** [comment]
bot, build com.moddb.TotalChaosRetro

File: com.moddb.TotalChaosRetro.yaml (hunk 5 above)
Comment (mbugni) [reply com.moddb.TotalChaosRetro.yaml]: @hfiguiere done, thanks @AsciiWolf

**Comment by AsciiWolf** [comment]
bot, build com.moddb.TotalChaosRetro

**Comment by AsciiWolf** [comment]
Wayland support looks to be broken with the older GZDoom version - the game runs in a small, borderless window and crashes a lot. Removing the wayland socket (and adding x11 one instead of fallback-x11) fixes it. Anyway, the game still suffers from strange flickering glitches when an OpenGL renderer is used:

glitch 1

glitch 2

**Comment by mbugni** [comment]
> Wayland support looks to be broken with the older GZDoom version - the game runs in a small, borderless window and crashes a lot. Removing the wayland socket (and adding x11 one instead of fallback-x11) fixes it. Anyway, the game still suffers from strange flickering glitches when an OpenGL renderer is used:
> 

Honestly I tested it with both KDE and GNOME (Fedora 37/38 and CentOS Stream 9) but always using X11, because I know many games are not wayland-friendly. I didn't know the x11 option and I added it, but the game in more or less the same on KDE.

@AsciiWolf what kind of GPU are you using?

Maybe we can add a warn to game description about X11 compatibility.

**Comment by mbugni** [comment]
> if it is buggy in wayland, disable wayland in the flatpak. it has to work out of the box.

Sorry @hfiguiere I'm new to flatpak. For disabling it at runtime it should be enough to add `--nosocket=wayland` while to remove it at build time I should remove the `--socket=wayland` from `finish-args`?

File: com.moddb.TotalChaosRetro.yaml [hunk 8]
```diff
@@ -0,0 +1,82 @@
+app-id: com.moddb.TotalChaosRetro
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
+sdk: org.freedesktop.Sdk
+command: gzdoom.sh
+finish-args:
+  - --device=all
+  - --socket=wayland
+  - --socket=fallback-x11
```
Comment (hfiguiere) [inline com.moddb.TotalChaosRetro.yaml:9]:
```suggestion
  - --socket=x11
```

since it's buggy with wayland

**Comment by AsciiWolf** [comment]
> @AsciiWolf what kind of GPU are you using?

Radeon RX 6600 XT.

**Comment by mbugni** [comment]
Just tested latest build using GNOME, both X11 and Wayland session. Works for me, both windowed and fullscreen mode (Intel Core i5-1035G1 UHD Graphics).

**Comment by AsciiWolf** [comment]
I wonder what causes these weird glitches when using OpenGL renderer in my case. They look like some kind of memory corruption.

**Comment by AsciiWolf** [comment]
It does.

**Comment by mbugni** [comment]
Can you try it with latest GZDoom engine? It should work, I suppose.

**Comment by AsciiWolf** [comment]
My Total Chaos Flatpak already uses more recent GZDoom version: https://github.com/flathub/com.moddb.TotalChaos/blob/master/com.moddb.TotalChaos.yaml#L49

You can try this version as well. It is confirmed to work fine with Total Chaos.

**Comment by mbugni** [comment]
I already tested the retro edition with latest gzdoom engine, it was my first build. For me both works, it's clearly a matter of GPU. I'd like to know if it works for you using the AMD driver.
BTW: I see there is gzdoom in shared-modules.

**Comment by AsciiWolf** [comment]
> BTW: I see there is gzdoom in shared-modules.

There is, however I have already discussed this with the Total Chaos developer and they told me that they tested Total Chaos with specific versions of GZDoom and don't guarantee that it will work with later ones, so I will rather keep GZDoom as a module directly in the Flatpak manifest.

**Comment by mbugni** [comment]
> You can try this version as well. It is confirmed to work fine with Total Chaos.

If you can confirm that the retro edition works using that engine, I'll switch to the same gzdoom version. I'll maintain the one with legacy engine as personal project, just for "old boys".  ;-)

**Comment by AsciiWolf** [comment]
You can update this Flatpak to use the same GZDoom version as my Total Chaos Flatpak. I will try the test build and let you know afterwards.

**Comment by mbugni** [comment]
bot, build com.moddb.TotalChaosRetro

**Comment by AsciiWolf** [comment]
Fullscreen works fine now, sadly there are still the same graphical glitches:

snow_glitch

**Comment by mbugni** [comment]
I don't know how to fix it. Did you play a bit with engine graphic options (hw acceleration, driver options etc)?
As far you know, authors are available for such kind of fixes?

I notice that latest build pulled a org.gtk.Gtk3theme.Breeze, but I don't know why.

**Comment by AsciiWolf** [comment]
Sorry, I also don't know how to fix it. I can give you contact to the Total Chaos developer via e-mail if you want.

**Comment by AsciiWolf** [comment]
It looks like that disabling Noise in Post Processing options did fix the issue. :-)

Everything else seems to work fine now.

Maybe the config file could be patched to disable Noise by default?

**Comment by mbugni** [comment]
Good news! As I mentioned you, I think the game runs much better by setting the following options:

- Post Processing
- Vignette > Enabled > No
- Noise > Enabled > No
- RETRO
- Enabled > No
- Enabled Pixelation > No
- Enabled Posterize > No

So a small patch to default config could help. @hfiguiere what do you think about?

**Comment by barthalion** [comment]
Sure, whatever makes it work out of the box.

**Comment by mbugni** [comment]
bot, build com.moddb.TotalChaosRetro

**Comment by mbugni** [comment]
@AsciiWolf can you test please the latest build?

**Comment by AsciiWolf** [comment]
I just did. Everything seems to work fine now. :-)

File: com.moddb.TotalChaosRetro.yaml (hunk 8 above)
Comment (mbugni) [reply com.moddb.TotalChaosRetro.yaml:9]: New GZDoom engine is not buggy as legacy one.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Total Chaos - Retro Edition
+
+This project contains files to build [Total Chaos - Retro Edition](https://www.moddb.com/mods/total-chaos/downloads/total-chaos-directors-cut-retro-edition-140) as flatpak app.
```
Comment (AsciiWolf) [inline README.md:3]:
```suggestion
This project contains files to build [Total Chaos - Retro Edition](https://www.moddb.com/mods/total-chaos/downloads/total-chaos-directors-cut-retro-edition-140) as a Flatpak app.
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+
+Retro Edition has reduced texture resolutions and poly-counts for improved performance. Save files from other versions of Total Chaos are not compatible.
```
Comment (AsciiWolf) [inline README.md:5]:
```suggestion
Retro Edition has reduced texture resolution and poly-counts for improved performance. Save files from other versions of Total Chaos are not compatible.
```

**Review by AsciiWolf — APPROVED** [review]

**Review by AsciiWolf — APPROVED** [review]
